web3.js/src/connection.ts

```diff
@@ -4263,75 +4207,273 @@ export class Connection {
       return;
     }
 
-    for (let clientSubscriptionId of accountClientSubscriptionIds) {
-      const sub = this._accountChangeSubscriptions[clientSubscriptionId];
-      this._subscribe(
-        sub,
-        'accountSubscribe',
-        this._buildArgs([sub.publicKey], sub.commitment, 'base64'),
-      );
-    }
-
-    for (let clientSubscriptionId of programClientSubscriptionIds) {
-      const sub = this._programAccountChangeSubscriptions[clientSubscriptionId];
-      this._subscribe(
-        sub,
-        'programSubscribe',
-        this._buildArgs([sub.programId], sub.commitment, 'base64', {
-          filters: sub.filters,
-        }),
-      );
-    }
-
-    for (let clientSubscriptionId of slotClientSubscriptionIds) {
-      const sub = this._slotSubscriptions[clientSubscriptionId];
-      this._subscribe(sub, 'slotSubscribe', []);
-    }
-
-    for (let clientSubscriptionId of slotUpdateClientSubscriptionIds) {
-      const sub = this._slotUpdateSubscriptions[clientSubscriptionId];
-      this._subscribe(sub, 'slotsUpdatesSubscribe', []);
-    }
-
-    for (let clientSubscriptionId of signatureClientSubscriptionIds) {
-      const sub = this._signatureSubscriptions[clientSubscriptionId];
-      const args: any[] = [sub.signature];
-      if (sub.options) args.push(sub.options);
-      this._subscribe(sub, 'signatureSubscribe', args);
-    }
+    await Promise.all(
+      Object.entries(
+        this._subscriptionsByHash as Record<
+          SubscriptionConfigHash,
+          Subscription
+        >,
+      ).map(async ([hash, subscription]) => {
+        switch (subscription.state) {
+          case 'pending':
+          case 'unsubscribed':
+            if (subscription.callbacks.size === 0) {
+              /**
+               * You can end up here when:
+               *
+               * - a subscription has recently unsubscribed
+               *   without having new callbacks added to it
+               *   while the unsubscribe was in flight, or
+               * - when a pending subscription has its
+               *   listeners removed before a request was
+               *   sent to the server.
+               *
+               * Being that nobody is interested in this
+               * subscription any longer, delete it.
+               */
+              delete this._subscriptionsByHash[hash];
+              if (subscription.state === 'unsubscribed') {
+                delete this._subscriptionCallbacksByServerSubscriptionId[
+                  subscription.serverSubscriptionId
+                ];
+              }
+              await this._updateSubscriptions();
+              return;
+            }
+            await (async () => {
```

Comment:
The subscription state could have changed away from `'pending'` or `'unsubscribed'` before this async function is scheduled. This also applies to how we handle unsubscribes.

Reply:
It can't, actually! Invoking an async function inline doesn't schedule it – it executes synchronously and eagerly.

```
async function sleep(ms) {
  return new Promise(resolve => {
    setTimeout(resolve, ms);
  });
}

console.log('A');
(async () => {
  console.log('synchronous state transition before B');
  await sleep(10);
  console.log('deferred code after continuation');
})();
console.log('B');
```

The same is true of Promises; the implementation executes eagerly and synchronously.

```
async function sleep(ms) {
  return new Promise(resolve => {
    setTimeout(resolve, ms);
  });
}

console.log('A');
new Promise(resolve => {
  console.log('synchronous state transition before B');
  sleep(10).then(() => {
    console.log('deferred code after continuation');
    resolve();
  });
});
console.log('B');
```

Reply:
Ah you're right, thanks